src/main/scala/quantiles/QuantileNonSample.scala

```diff
@@ -0,0 +1,252 @@
+package quantiles
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.reflect.ClassTag
+import scala.util.control.Breaks._
+import scala.util.Sorting.quickSort
+
+class QuantileNonSample[T](val sketchSize: Int,
+                           private val shrinkingFactor: Double = 0.64)
+                          (implicit ordering: Ordering[T],
+                           ct: ClassTag[T]) extends Serializable{
+
+  private var curNumOfCompactors = 0
+  // initialize with ArrayBuffer, add compactors later
+  private var compactors = ArrayBuffer[NonSampleCompactor[T]]()
+  // number of items in compactors
+  private var compactorActualSize = 0
+  // overall capacity of compactors
+  private var compactorTotalSize = 0
+  expand()
+
+
+
+  // expand a layer of compactor
+  private def expand(): Unit = {
+    compactors = compactors :+ new NonSampleCompactor[T]
+    curNumOfCompactors = compactors.length
+    var size = 0
+    (0 until curNumOfCompactors).foreach { height =>
+      size = size + capacity(height)
+    }
+    compactorTotalSize = size
+  }
+
+  private def capacity(height:Int): Int = {
+    Math.ceil(sketchSize * Math.pow(shrinkingFactor, height)).toInt + 1
+  }
+
+  /**
+    * update the sketch with a single item
+    *
+    * @param item new item observed by the sketch
+    */
+  def update(item: T): Unit = {
+    compactors(0).buffer = compactors(0).buffer :+ item
+
+    compactorActualSize = compactorActualSize + 1
+    if (compactorActualSize > compactorTotalSize) {
+      condense()
+    }
+  }
+
+  private def condense(): Unit = {
+    breakable {
+      for (height <- compactors.indices) {
+        if (compactors(height).buffer.length >= capacity(height)) {
+          if (height + 1 >= curNumOfCompactors) expand()
+          val output: Array[T] = compactors(height).compact
+          output.foreach(element => compactors(height + 1).buffer = compactors(height + 1).buffer :+ element)
+          var size = 0
+          (0 until curNumOfCompactors).foreach { height =>
+            size = size + this.compactors(height).buffer.length
+          }
+          compactorActualSize = size
+          // implement lazy here
+          break
+        }
+      }
+    }
+  }
+
+  private def compare[T](o1: T, o2: T)(implicit ord: Ordering[T]): Boolean = ord.gt(o1, o2)
+
+  /**
+    * Get the map which contains the rank of all items which is currently in the sketch.
+    * @return the map which contains the rank of all items which is currently in the sketch (RankMap)
+    */
+  def getRankMap(): mutable.Map[T,Long] = {
```

Comment:
The double loop on output can work but isn't the right complexity. You need a loop over the items of "output" sorted by their value (tuple._1). When visiting an item, add its weight to the running weight and add to the mutable map (states): states(item) = running_weight. Notice that the assignment needs to happen after adding the item's weight to the running weight.

If you do this, ignore the other two comments for getRankMap

Reply:
Yes I have changed the implementation to your optimized design.